src/clients/c++/simple_shm_client.cc

```diff
@@ -0,0 +1,340 @@
+    exit(1);
+  }
+
+  ni::ServerStatus server_status;
+  err = status_ctx->GetServerStatus(&server_status);
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to get status: " << err << std::endl;
+    exit(1);
+  }
+
+  std::cout << "Status for model " << model_name << ":" << std::endl;
+  std::cout << server_status.DebugString() << std::endl;
+
+  // Create the inference context for the model.
+  std::unique_ptr<nic::InferContext> infer_ctx;
+  if (protocol == "http") {
+    err = nic::InferHttpContext::Create(
+        &infer_ctx, url, http_headers, model_name, -1 /* model_version */,
+        verbose);
+  } else if (protocol == "grpc") {
+    err = nic::InferGrpcContext::Create(
+        &infer_ctx, url, model_name, -1 /* model_version */, verbose);
+  } else {
+    Usage(argv, "unknown protocol '" + protocol + "'");
+  }
+
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to create inference context: " << err
+              << std::endl;
+    exit(1);
+  }
+
+  // Set the context options to do batch-size 1 requests. Also request that
+  // all output tensors be returned.
+  std::unique_ptr<nic::InferContext::Options> options;
+  FAIL_IF_ERR(
+      nic::InferContext::Options::Create(&options),
+      "unable to create inference options");
+
+  options->SetBatchSize(1);
+  FAIL_IF_ERR(
+      infer_ctx->SetRunOptions(*options), "unable to set inference options");
+
+  std::shared_ptr<nic::InferContext::Input> input0, input1;
+  std::shared_ptr<nic::InferContext::Output> output0, output1;
+  FAIL_IF_ERR(infer_ctx->GetInput("INPUT0", &input0), "unable to get INPUT0");
+  FAIL_IF_ERR(infer_ctx->GetInput("INPUT1", &input1), "unable to get INPUT1");
+  FAIL_IF_ERR(
+      infer_ctx->GetOutput("OUTPUT0", &output0), "unable to get OUTPUT0");
+  FAIL_IF_ERR(
+      infer_ctx->GetOutput("OUTPUT1", &output1), "unable to get OUTPUT1");
+
+  FAIL_IF_ERR(input0->Reset(), "unable to reset INPUT0");
+  FAIL_IF_ERR(input1->Reset(), "unable to reset INPUT1");
+
+  size_t input_byte_size = 16 * sizeof(int32_t);
+  size_t output_byte_size = 16 * sizeof(int32_t);
+
+  // Create Input0 and Input1 in Shared Memory. Initialize Input0 to unique
+  // integers and Input1 to all ones.
+  std::string shm_key = "/input_simple";
+  int shm_fd = create_shared_region(shm_key, input_byte_size * 2);
+  int* input0_shm = (int*)(get_shm_addr(shm_fd, 0, input_byte_size * 2));
+  int* input1_shm = (int*)(input0_shm + 16);
+  for (size_t i = 0; i < 16; ++i) {
+    *(input0_shm + i) = i;
+    *(input1_shm + i) = 1;
+  }
+
+  // Create Output0 and Output1 in Shared Memory
+  shm_key = "/output_simple";
+  shm_fd = create_shared_region(shm_key, output_byte_size * 2);
+  int* output0_shm = (int*)(get_shm_addr(shm_fd, 0, output_byte_size * 2));
+  int* output1_shm = (int*)(output0_shm + 16);
+
+  // Register all Inputs and Outputs with TRTIS
+  infer_ctx->RegisterSharedMemory("input0_data", "/input_simple", 0, input_byte_size);
```

Comment:
Just register 1 shared memory region for inputs and 1 for outputs. Then in SetSharedMemory use offsets.